File: demo/quant_trading_system/src/quant_trading_system/api/routers/backtest.py
"""
回测路由
========

回测API接口
"""

from typing import Any

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

router = APIRouter()


class BacktestRequest(BaseModel):
    """回测请求"""
    strategy_type: str
    symbol: str
    timeframe: str = "1m"
    start_time: str  # ISO格式时间
    end_time: str
    initial_capital: float = 1000000.0
    commission_rate: float = 0.0004
    slippage_rate: float = 0.0001
    params: dict[str, Any] = {}


@router.post("/run")
async def run_backtest(request: BacktestRequest) -> dict[str, Any]:
    """运行回测"""
    return {
        "success": True,
        "backtest_id": "new_backtest_id",
        "message": "Backtest started",
    }


@router.get("/{backtest_id}")
async def get_backtest_result(backtest_id: str) -> dict[str, Any]:
    """获取回测结果"""
    return {
        "backtest_id": backtest_id,
        "status": "completed",
        "result": {
            "total_return": 0,
            "annual_return": 0,
            "max_drawdown": 0,
            "sharpe_ratio": 0,
            "total_trades": 0,
            "win_rate": 0,
        }
    }


@router.get("/{backtest_id}/equity")
async def get_backtest_equity(backtest_id: str) -> dict[str, Any]:
    """获取回测权益曲线"""
    return {
        "backtest_id": backtest_id,
        "equity_curve": [],
    }


@router.get("/{backtest_id}/trades")
async def get_backtest_trades(
    backtest_id: str,
    limit: int = Query(100, ge=1, le=1000),
) -> dict[str, Any]:
    """获取回测交易记录"""
    return {
        "backtest_id": backtest_id,
        "trades": [],
        "total": 0,
    }


@router.get("/list")
async def list_backtests(
    limit: int = Query(20, ge=1, le=100),
) -> dict[str, Any]:
    """获取回测历史列表"""
    return {
        "backtests": [],
        "total": 0,
    }


@router.delete("/{backtest_id}")
async def delete_backtest(backtest_id: str) -> dict[str, Any]:
    """删除回测记录"""
    return {
        "success": True,
        "backtest_id": backtest_id,
        "message": "Backtest deleted",
    }

File: web_site/src/types/api/backtest.ts
/**
 * 回测相关API类型定义
 */

// ==================== 回测类型 ====================

/**
 * 回测状态
 */
export type BacktestStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'

/**
 * 运行回测请求
 */
export interface RunBacktestRequest {
  strategy_type: string
  symbol: string
  timeframe?: string
  start_time: string
  end_time: string
  initial_capital?: number
  commission_rate?: number
  slippage_rate?: number
  params?: Record<string, any>
}

/**
 * 回测结果
 */
export interface BacktestResult {
  id: string
  user_id: string
  strategy_id: string
  strategy_name?: string
  exchange_id: string
  exchange_name?: string
  symbol: string
  status: BacktestStatus
  start_time: string
  end_time: string
  initial_capital: number
  final_capital: number
  total_return: number
  total_return_ratio: number
  annual_return_ratio: number
  max_drawdown: number
  max_drawdown_ratio: number
  sharpe_ratio: number
  sortino_ratio: number
  calmar_ratio: number
  win_rate: number
  profit_factor: number
  total_trades: number
  winning_trades: number
  losing_trades: number
  avg_profit: number
  avg_loss: number
  max_profit: number
  max_loss: number
  avg_holding_time: number
  commission_paid: number
  parameters?: Record<string, any>
  error_message?: string
  run_time?: number
  create_time: string
  complete_time?: string
}

/**
 * 回测历史列表查询参数
 */
export interface BacktestListParams {
  limit?: number
}

/**
 * 回测历史列表响应
 */
export interface BacktestListResponse {
  backtests: BacktestResult[]
  total: number
}

/**
 * 权益曲线数据点
 */
export interface EquityPoint {
  time: string
  equity: number
  cash: number
  position_value: number
  return_ratio: number
  drawdown: number
  drawdown_ratio: number
}

/**
 * 权益曲线响应
 */
export interface EquityCurveResponse {
  backtest_id: string
  initial_capital: number
  final_capital: number
  data: EquityPoint[]
}

/**
 * 回测交易记录
 */
export interface BacktestTrade {
  id: string
  backtest_id: string
  symbol: string
  side: 'buy' | 'sell'
  type: 'open' | 'close'
  price: number
  quantity: number
  commission: number
  pnl?: number
  pnl_ratio?: number
  balance_after: number
  signal_reason?: string
  trade_time: string
}

/**
 * 回测交易记录查询参数
 */
export interface BacktestTradesParams {
  limit?: number
}

/**
 * 回测交易记录响应
 */
export interface BacktestTradesResponse {
  backtest_id: string
  trades: BacktestTrade[]
  total: number
}
